# Tratamento de dados 

In [1]:
import pandas as pd

In [2]:
dados = pd.read_csv('dados_residencial_final.csv', sep=';')

In [3]:
dados.head(10)

Tipo           Bairro  Quartos  Vagas  Suites  Area    Valor  \
0           Quitinete       Copacabana        1      0       0    40   1700.0   
1                Casa  Jardim Botânico        2      0       1   100   7000.0   
2         Apartamento           Centro        1      0       0    15    800.0   
3         Apartamento     Higienópolis        1      0       0    48    800.0   
4         Apartamento     Vista Alegre        3      1       0    70   1200.0   
5         Apartamento         Cachambi        2      0       0    50   1300.0   
6  Casa de Condomínio  Barra da Tijuca        5      4       5   750  22000.0   
7  Casa de Condomínio            Ramos        2      2       0    65   1000.0   
8         Apartamento           Centro        1      0       0    36   1200.0   
9         Apartamento           Grajaú        2      1       0    70   1500.0   

   Condominio  IPTU  
0       500.0  60.0  
1         NaN   NaN  
2       390.0  20.0  
3       230.0   NaN  
4         NaN   NaN  
5       301.0  17.0  
6         NaN   NaN  
7         NaN   NaN  
8         NaN   NaN  
9       642.0  74.0

In [4]:
dados.isnull()

Tipo  Bairro  Quartos  Vagas  Suites   Area  Valor  Condominio   IPTU
0      False   False    False  False   False  False  False       False  False
1      False   False    False  False   False  False  False        True   True
2      False   False    False  False   False  False  False       False  False
3      False   False    False  False   False  False  False       False   True
4      False   False    False  False   False  False  False        True   True
...      ...     ...      ...    ...     ...    ...    ...         ...    ...
23061  False   False    False  False   False  False  False       False  False
23062  False   False    False  False   False  False  False       False  False
23063  False   False    False  False   False  False  False       False  False
23064  False   False    False  False   False  False  False       False  False
23065  False   False    False  False   False  False  False       False   True

[23066 rows x 9 columns]

In [6]:
dados.notnull()

Tipo  Bairro  Quartos  Vagas  Suites  Area  Valor  Condominio   IPTU
0      True    True     True   True    True  True   True        True   True
1      True    True     True   True    True  True   True       False  False
2      True    True     True   True    True  True   True        True   True
3      True    True     True   True    True  True   True        True  False
4      True    True     True   True    True  True   True       False  False
...     ...     ...      ...    ...     ...   ...    ...         ...    ...
23061  True    True     True   True    True  True   True        True   True
23062  True    True     True   True    True  True   True        True   True
23063  True    True     True   True    True  True   True        True   True
23064  True    True     True   True    True  True   True        True   True
23065  True    True     True   True    True  True   True        True  False

[23066 rows x 9 columns]

## Identificando valores nullos usando .info()

In [9]:
dados.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 23066 entries, 0 to 23065
Data columns (total 9 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   Tipo        23066 non-null  object 
 1   Bairro      23066 non-null  object 
 2   Quartos     23066 non-null  int64  
 3   Vagas       23066 non-null  int64  
 4   Suites      23066 non-null  int64  
 5   Area        23066 non-null  int64  
 6   Valor       23057 non-null  float64
 7   Condominio  21198 non-null  float64
 8   IPTU        16196 non-null  float64
dtypes: float64(3), int64(4), object(2)
memory usage: 1.6+ MB


#### para visualizar, por exemplo, 
#### os valores nulos dentro do campo 'Valor'

In [10]:
dados[dados['Valor'].isnull()]

Tipo           Bairro  Quartos  Vagas  Suites  Area  Valor  \
60    Apartamento  Barra da Tijuca        2      1       1    70    NaN   
1521  Apartamento             Leme        2      0       0    75    NaN   
1715         Casa     Campo Grande        3      4       3   363    NaN   
2052  Apartamento         Botafogo        2      0       0    95    NaN   
2074  Apartamento       Copacabana        2      0       0    72    NaN   
5049         Casa     Campo Grande        3      2       1   100    NaN   
8756  Apartamento             Leme        2      0       1    75    NaN   
9144  Apartamento           Glória        3      0       1   135    NaN   
9349  Apartamento            Gávea        3      1       1   105    NaN   

      Condominio   IPTU  
60         970.0   68.0  
1521       878.0    NaN  
1715         NaN    NaN  
2052      1010.0  170.0  
2074       850.0    NaN  
5049         NaN    NaN  
8756       878.0    NaN  
9144       910.0  228.0  
9349       880.0  221.0

#### removendo valores nulos 

A = dados.shape[0]
dados.dropna(subset = ['Valor'], inplace = True)
B = dados.shape[0]
A - B

In [12]:
dados[dados['Valor'].isnull()]

Empty DataFrame
Columns: [Tipo, Bairro, Quartos, Vagas, Suites, Area, Valor, Condominio, IPTU]
Index: []

#### dataframe vazio depois do .dropna()

#### tratamento para Condominio, por exemplo, não é tão simples quanto 
#### simplesmente dropar os valores nulos, casas não possuem condimínio, então será preciso analisar melhor os dados

## Tratamento da coluna Condomínio

#### para descobrir os valores nulos dentro do field

In [13]:
dados[dados['Condominio'].isnull()].shape[0]


1866

#### verificando quais desses são em apartamentos: 

In [17]:
selecao = (dados['Tipo'] == 'Apartamento') & (dados['Condominio'].isnull())

##### Contudo, se nos atentarmos a nossa seleção, verificaremos que ela nos fornece exatamente as assinaturas que deverão ser removidas do banco de dados. Reaproveitaremos a estrutura já utilizada anteriormente neste notebook, e passaremos a seleção que criamos (selecao). Para eliminar as assinaturas nulas selecionadas ao invés de coletá-las, iremos adicionar ~ em selecao, que inverte a Series booleana. 

In [18]:
A = dados.shape[0] #matriz original
dados = dados[~selecao] #removendo os valores selecionados anterioremente
B = dados.shape[0]
A - B

745

In [19]:
dados[dados['Condominio'].isnull()].shape[0]

1121

###### os valores restantes de 1121 são aqueles que tem condomínio nulo, mas não são apartamentos

In [20]:
dados.fillna(0, inplace = True)

/var/folders/0h/x896y6yn107gp3wxmw7g6rk80000gp/T/ipykernel_1676/598842260.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  dados.fillna(0, inplace = True)


In [21]:
dados.head(10)

Tipo               Bairro  Quartos  Vagas  Suites  Area  \
0            Quitinete           Copacabana        1      0       0    40   
1                 Casa      Jardim Botânico        2      0       1   100   
2          Apartamento               Centro        1      0       0    15   
3          Apartamento         Higienópolis        1      0       0    48   
5          Apartamento             Cachambi        2      0       0    50   
6   Casa de Condomínio      Barra da Tijuca        5      4       5   750   
7   Casa de Condomínio                Ramos        2      2       0    65   
9          Apartamento               Grajaú        2      1       0    70   
10         Apartamento  Lins de Vasconcelos        3      1       1    90   
11         Apartamento           Copacabana        1      0       1    40   

      Valor  Condominio  IPTU  
0    1700.0       500.0  60.0  
1    7000.0         0.0   0.0  
2     800.0       390.0  20.0  
3     800.0       230.0   0.0  
5    1300.0       301.0  17.0  
6   22000.0         0.0   0.0  
7    1000.0         0.0   0.0  
9    1500.0       642.0  74.0  
10   1500.0       455.0  14.0  
11   2000.0       561.0  50.0

###### outra maneira de fazer a mesma substituição

In [22]:
dados = dados.fillna({'Condominio': 0, 'IPTU': 0}) # aqui é mais favorável para definir valores diferentes

In [23]:
dados.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 22312 entries, 0 to 23065
Data columns (total 9 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   Tipo        22312 non-null  object 
 1   Bairro      22312 non-null  object 
 2   Quartos     22312 non-null  int64  
 3   Vagas       22312 non-null  int64  
 4   Suites      22312 non-null  int64  
 5   Area        22312 non-null  int64  
 6   Valor       22312 non-null  float64
 7   Condominio  22312 non-null  float64
 8   IPTU        22312 non-null  float64
dtypes: float64(3), int64(4), object(2)
memory usage: 1.7+ MB


In [24]:
# novo DataFrame sem nenhum valor nulo

In [25]:
# melhorando novamente nosso dataFrame: 

In [27]:
dados.to_csv('dados_residencial_final_v2.csv', sep=';', index = False)